### PR TITLE
[streaming]Add master and scheduler.

### PR DESCRIPTION
**Pull request by BalaBalaYi**
This PR mainly includes the implementation of master(JobMaster) part.

1. Add JobMaster: 
    1) Init GraphManager, ResourceManager, JobScheduler and do scheduling when job is submitted.

2. Add JobScheduler:
    1) Use ResourceManager to allocate resource according to the graph info.
    2) Use WorkerLifecycleController to create, init, start and destroy workers.

3. Add WorkerLifecycleController:
    1) Create worker actors.
    2) Init worker actors.
    3) Start worker actors.
    4) Destroy worker actors.

4. JobWorkerContext: 
    1) Necessary info for JobWorker to run a task.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    this.conf = streamingConfig.masterConfig;
+
+    // init runtime context
+    runtimeContext = new JobRuntimeContext(streamingConfig);
+
+    LOG.info("Job master init success.");
+  }
+
+  /**
+   * Init JobMaster.
```
Comment (chaokunyang) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java:48]: Need more document

**Review by chaokunyang — APPROVED** [review]
LGTM, thanks

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/config/Config.java [hunk 2]
```diff
@@ -1,11 +1,10 @@
...
 
-import java.io.Serializable;
-import javax.accessibility.Accessible;
+import org.aeonbits.owner.Accessible;
 
 /**
  * Basic config interface.
  */
-public interface Config extends org.aeonbits.owner.Config, Accessible, Serializable {
+public interface Config extends org.aeonbits.owner.Config, Accessible {
```
Comment (datayjz) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/config/Config.java:8]: Why remove ```Serializable```, configuration needs  to be passed across the network.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/config/master/SchedulerConfig.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  /**
+   * The timeout ms of worker starting.
+   * Default is: 10000ms(10s).
+   *
+   * @return timeout ms
+   */
+  @Key(WORKER_STARTING_WAIT_TIMEOUT_MS)
+  @DefaultValue(value = "10000")
+  int workerStartingWaitTimeoutMs();
```
Comment (datayjz) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/config/master/SchedulerConfig.java:31]: What is different between ```Initiation``` and ```Starting```? It is better to rename ```Creating``` to  ```Starting```?

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    this.jobMasterActor = jobMasterActor;
+
+    // init manager
+    graphManager = new GraphManagerImpl(runtimeContext);
+    resourceManager = new ResourceManagerImpl(runtimeContext);
+
+    // build and set graph into runtime context
+    ExecutionGraph executionGraph = graphManager.buildExecutionGraph(jobGraph);
+    runtimeContext.setExecutionGraph(executionGraph);
```
Comment (datayjz) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java:96]: ```JobGraph``` also needs to give ```runtimeContext```.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    Language language = executionVertex.getLanguage();
+
+    ActorCreationOptions options = new ActorCreationOptions.Builder()
+        .setResources(executionVertex.getResources())
+        .setMaxReconstructions(ActorCreationOptions.INFINITE_RECONSTRUCTION)
+        .createActorCreationOptions();
+
+    RayActor<JobWorker> actor = null;
+    if (Language.JAVA == language) {
+      // TODO (yunye)
```
Comment (datayjz) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java:55]:
```suggestion
      // TODO (datayjz)
```

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    } else {
+      // TODO (yunye)
```
Comment (datayjz) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java:57]:
```suggestion
      // TODO (datayjz)
```

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/config/Config.java (hunk 2 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/config/Config.java:8]: Owner config already implements ```Serializable ```.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/config/master/SchedulerConfig.java (hunk 3 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/config/master/SchedulerConfig.java:31]: Can not use ray.wait() on actor creating.

Initiation -> Use wait() to update worker's context.
Starting -> Use wait() to start worker's task.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java (hunk 4 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java:96]: Fixed.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    if (this.runtimeContext.getExecutionGraph() == null) {
+      LOG.error("Init job master failed. Job graphs is null.");
+      return false;
+    }
+
+    // recover from last checkpoint
+    if (isRecover) {
+      // TODO (lianxin)
+    }
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java:64]: Complete this TODO or remove this if block.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+    ExecutionGraph executionGraph = graphManager.getExecutionGraph();
+    Preconditions.checkArgument(executionGraph != null, "no execution graph");
+
+    LOG.info("Finish to init job master.");
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java:69]:
```suggestion
    LOG.info("Finished initializing job master.");
```

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/JobScheduler.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package io.ray.streaming.runtime.master.scheduler;
+
+import io.ray.streaming.runtime.core.graph.executiongraph.ExecutionGraph;
+
+/**
+ * Job scheduler is used to do the schedule job in JobMaster.
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/JobScheduler.java:6]:
```suggestion
 * Job scheduler is used to do the scheduling job in JobMaster.
```

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/JobSchedulerImpl.java [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    this.resourceManager = jobMaster.getResourceManager();
+    this.workerLifecycleController = new WorkerLifecycleController();
+    this.jobConf = jobMaster.getRuntimeContext().getConf();
+
+    LOG.info("Scheduler init success.");
+  }
+
+  @Override
+  public boolean scheduleJob(ExecutionGraph executionGraph) {
+    LOG.info("Start to schedule job: {}.", executionGraph.getJobName());
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/JobSchedulerImpl.java:46]:
```suggestion
    LOG.info("Started scheduling job: {}.", executionGraph.getJobName());
```

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        .setResources(executionVertex.getResources())
+        .setMaxReconstructions(ActorCreationOptions.INFINITE_RECONSTRUCTION)
+        .createActorCreationOptions();
+
+    RayActor<JobWorker> actor = null;
+    if (Language.JAVA == language) {
+      // TODO (datayjz)
+    } else {
+      // TODO (datayjz)
+    }
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java:58]: same here

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    executionGraph.getNonSourceActors()
+        .forEach(actor -> rayObjects.add(RemoteCallWorker.startWorker(actor)));
+
+    WaitResult<Boolean> result = Ray.wait(rayObjects, rayObjects.size(), timeout);
+    if (result.getReady().size() != rayObjects.size()) {
+      LOG.error("Start workers timeout[{} ms].", timeout);
+      return false;
+    }
+
+    LOG.info("Start workers succeeded. Cost {} ms.", System.currentTimeMillis() - startTime);
```
Comment (raulchen) [inline streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java:131]:
```suggestion
    LOG.info("Workers started. Costed {} ms.", System.currentTimeMillis() - startTime);
```

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java (hunk 7 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/JobMaster.java:64]: Remove 'recover' for now. Will add back in 'checkpoint' part.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/JobScheduler.java (hunk 9 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/JobScheduler.java:6]: Job scheduler is used to do the scheduling in JobMaster.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java (hunk 12 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java:131]: Same modification on create, init and destroy.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java (hunk 11 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/controller/WorkerLifecycleController.java:58]: 'if' block removed.

File: streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/JobSchedulerImpl.java (hunk 10 above)
Comment (BalaBalaYi) [reply streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/master/scheduler/JobSchedulerImpl.java:46]: Use 'Begin scheduling'.

**Pull request merged**